python/lsst/dax/distribution/chunklogs.py

```diff
@@ -0,0 +1,413 @@
+        """
+        # problem_set is the set of all chunks that should be checked
+        # by hand before being generated again. Once checked, problem
+        # chunk Ids should be removed from assigned and limbo logs.
+        problem_set = self._assigned.chunk_set.difference(self._completed.chunk_set)
+        problem_set.update(self._limbo.chunk_set)
+        notstarted_set = self._target.chunk_set.difference(self._completed.chunk_set)
+        notstarted_set.difference_update(problem_set)
+        rpt = f'Problem chunk ids:\n{problem_set}\n\n'
+        rpt += f'Log counts:\n'
+        rpt += f' Target:     {len(self._target.chunk_set)}\n'
+        rpt += f' Assigned:   {len(self._assigned.chunk_set)}\n'
+        rpt += f' Completed:  {len(self._completed.chunk_set)}\n'
+        rpt += f' Limbo:      {len(self._limbo.chunk_set)}\n\n'
+        rpt += f' Problem:    {len(problem_set)}\n'
+        rpt += f' not started:{len(notstarted_set)}'
+        return rpt
+
+    def build(self, all_valid_chunks):
+        """Build this object from paramters and files.
+
+        Parameters
+        ----------
+        all_valid_chunks : list of int
+            List of all valid chunks, used to remove invalid chunk ids
+            from target.
+
+        Note
+        ------
+        self.result_set is produced here. it is a set of target chunks with
+        all invalid, completed, assigned, and limbo chunks removed.
+        """
+
+        #Handle raw string input
+        raw_in = None
+        if self._target_raw:
+            raw_in = ChunkListFile(None)
+            # Use comma for separator since this came from the command line.
+            raw_in.parse(self._target_raw, ',')
+
+        if self._target._fname:
+            # Use the file, if provided. If raw text input
+            # was provided, use the intersection of that and the
+            # target file.
+            self._target.read()
+            if raw_in:
+                self._target.chunk_set.intersection_update(raw_in.chunk_set)
+        elif raw_in:
+            # Use provided raw string
+            self._target = raw_in
+        else:
+            # Target is all valid chunks
+            self._target.chunk_set = set(all_valid_chunks)
+        # Make sure no invalid chunks are in the ltarget set.
+        if all_valid_chunks is not None:
+            self._target.intersectWithValid(all_valid_chunks)
+
+        # result_set is the target set with all chunks found in
+        # complete, assigned, and limbo removed. Technically
+        # this could just be target - assigned, but users
+        # are expected to edit the files and ingesting the
+        # same chunk twice could be bad.
+        self.result_set = self._target.chunk_set.copy()
+
+        # Remove chunks from result_set if they are in completed,
+        # assigned, or limbo.
+        lst = [self._completed, self._assigned, self._limbo]
+        for item in lst:
+            if item._fname:
```

Comment:
Isn't _fname supposed to be private?

Reply:
In C++ I would have made ChunkLogs a friend of ChunkListFile, as they are closely related classes. I don't really want anything else (at least outside this file) looking at _fname. I'm not sure how to set that up with python.